hw/darwin/quartz/XDarwinStartup.man
.TH XDarwinStartup 1
.SH NAME
XDarwinStartup - Startup program for the XDarwin X window server
.SH SYNOPSIS
.B XDarwinStartup
[\fI-iokit\fP]
[\fI-fullscreen\fP]
[\fI-rootless\fP]
[\fI-quartz\fP]
[\fI-idle\fP]
[\fIoptions\fP]
.SH DESCRIPTION
The \fIXDarwin(1)\fP X window server can be run in a variety of different
modes and is actually two different executables. The IOKit X server,
XDarwin, is used when running from the console. It is most commonly
located in __XBinDir__. The Quartz X server, for running in parallel with
Aqua, is a full-fledged Mac OS X application that can be started from
the Finder. Its application bundle is XDarwin.app, which is typically
located in /Applications.
.I XDarwinStartup
allows easy switching between these X servers and auto-detection of the
appropriate one to use when launching from the command line.
When run without any arguments,
.I XDarwinStartup
will start the Quartz X server if the Core Graphics window server
is currently running. Otherwise it will start the IOKit X server.
.PP
To locate the Quartz X server,
.I XDarwinStartup
will try to read the soft link at __XBinDir__/XDarwinQuartz.
This is typically a soft link to the executable of the XDarwin.app
application. If this fails,
.I XDarwinStartup
will call Launch Services to locate XDarwin.app.
.PP
To start the IOKit X server,
.I XDarwinStartup
will run the XDarwin executable, which should be present in the
user's path.
.SH OPTIONS
.I XDarwinStartup
accepts and passes on all options to the X server it
launches. In addition the following options have specific effects:
.TP 8
.B \-iokit
Launch the IOKit X server.
.TP 8
.B \-fullscreen
Launch the Quartz X server to run in full screen mode.
.TP 8
.B \-rootless
Launch the Quartz X server to run in rootless mode.
.TP 8
.B \-quartz
Launch the Quartz X server.
.TP 8
.B \-idle
Pause and do nothing. This option is used by XDarwin.app when it is
started from the Finder.
.SH FILES
.TP 30
__XBinDir__/XDarwin
IOKit mode X server
.TP 30
/Applications/XDarwin.app
Quartz mode X server
.TP 30
__XBinDir__/XDarwinQuartz
Soft link to Quartz mode X server executable
.SH SEE ALSO 
XDarwin(1)
.SH BUGS
The path to XDarwinQuartz should not be hard coded. 



ChangeLog
2006-06-05  Daniel Stone  <daniel@freedesktop.org>

	* hw/xfree86/os-support/bus/Pci.c:
	* hw/xfree86/os-support/bus/Pci.h:
	* hw/xfree86/os-support/bus/linuxPci.c:
	* hw/xfree86/os-support/linux/lnx_pci.c:
	When we can, bound the maximum number of PCI devices to attempt to
	scan, by the number found on the system.  Only implemented for Linux
	right now.

2006-06-05  Keith Packard  <keithp@keithp.com>

	* acinclude.m4:
	Update AC_DEFINE_DIR to version compatible with CVS autoconf
	which requires double eval to avoid leaving ${prefix} in output

2006-06-04  Adam Jackson  <ajax@freedesktop.org>

	* hw/xfree86/loader/dixsym.c:
	Remove a stray LBX reference.

	* hw/xfree86/ddc/xf86DDC.c:
	* hw/xfree86/dixmods/GLcoremodule.c:
	* hw/xfree86/dixmods/afbmodule.c:
	* hw/xfree86/dixmods/cfb32module.c:
	* hw/xfree86/dixmods/cfbmodule.c:
	* hw/xfree86/dixmods/dbemodule.c:
	* hw/xfree86/dixmods/fbmodule.c:
	* hw/xfree86/dixmods/ftmodule.c:
	* hw/xfree86/dixmods/glxmodule.c:
	* hw/xfree86/dixmods/mfbmodule.c:
	* hw/xfree86/dixmods/recordmod.c:
	* hw/xfree86/dixmods/shmodule.c:
	* hw/xfree86/dixmods/type1mod.c:
	* hw/xfree86/dixmods/xf86XTrapModule.c:
	* hw/xfree86/dixmods/extmod/modinit.c:
	* hw/xfree86/dri/drimodule.c:
	* hw/xfree86/exa/examodule.c:
	* hw/xfree86/fbdevhw/fbdevhw.c:
	* hw/xfree86/i2c/bt829_module.c:
	* hw/xfree86/i2c/fi1236_module.c:
	* hw/xfree86/i2c/msp3430_module.c:
	* hw/xfree86/i2c/tda8425_module.c:
	* hw/xfree86/i2c/tda9850_module.c:
	* hw/xfree86/i2c/tda9885_module.c:
	* hw/xfree86/i2c/uda1380_module.c:
	* hw/xfree86/i2c/xf86i2cmodule.c:
	* hw/xfree86/int10/xf86int10module.c:
	* hw/xfree86/os-support/drm/drmmodule.c:
	* hw/xfree86/rac/xf86RACmodule.c:
	* hw/xfree86/ramdac/xf86RamDacMod.c:
	* hw/xfree86/scanpci/xf86ScanPci.c:
	* hw/xfree86/shadowfb/sfbmodule.c:
	* hw/xfree86/vbe/vbe_module.c:
	* hw/xfree86/vgahw/vgaHWmodule.c:
	* hw/xfree86/xaa/xaaInitAccel.c:
	* hw/xfree86/xf1bpp/mfbmodule.c:
	* hw/xfree86/xf4bpp/vgamodule.c:
	* hw/xfree86/xf8_16bpp/xf8_16module.c:
	* hw/xfree86/xf8_32bpp/cfb8_32module.c:
	Ensure all *ModuleData symbols are marked _X_EXPORT.  Start removing
	XFree86LOADER ifdefs, non-loadable hasn't been supported for a while
	now.  Remove completely gratuitious REMOVE_LOADER_CHECK_MODULE_INFO
	ifdefs surrounding a call to a function added in XFree86 4.1 (!).
	Miscellaneous static markings.

2006-06-03  Daniel Stone  <daniel@freedesktop.org>

	* configure.ac:
	Bump to 1.1.99.2.

	* configure.ac:
	* hw/xfree86/os-support/bus/Pci.h:
	Add support for kFreeBSD systems.  (Robert Millan, Aurelien Jarno)

	* hw/xfree86/loader/loader.h:
	* hw/xfree86/loader/loader.c:
	Completely remove relocation pointer tables.

	* Xext/appgroup.c:
	Bug #6956: Fix crash when removing session leader before its children.
	(Rich Coe)

	* hw/xwin/wincursor.c:
	* hw/xwin/winresource.h:
	* hw/xwin/winwndproc.c:
	* hw/xwin/XWin.rc:
	Bug #6619: Fix disappearing hardware cursor.  (Colin Harrison)

2006-06-01  Adam Jackson  <ajax@freedesktop.org>

	* hw/xfree86/common/xf86Init.c:
	* hw/xfree86/dixmods/Makefile.am:
	* hw/xfree86/dixmods/bitmapmod.c:
	Bug #5089: Die, libbitmap, die!

	* hw/xfree86/common/xf86Config.c:
	Add bitmap to the ignored module list.

2006-06-01  Daniel Stone  <daniel@freedesktop.org>

	* Xext/Makefile.am:
	* Xext/README.xtest1-ddx:
	* Xext/xtest1di.c:
	* Xext/xtest1dd.h:
	* Xext/xtest1dd.c:
	* Xprint/ddxInit.c:
	* hw/kdrive/src/kdrive.c:
	* hw/xfree86/common/xf86Events.c:
	* hw/xfree86/common/xf86Init.c:
	* hw/xfree86/os-support/solaris/sun_kbdEv.c:
	* hw/xnest/TestExt.c:
	* hw/xwin/InitInput.c:
	* hw/vfb/InitInput.c:
	* mi/miinitext.c:
	* os/WaitFor.c:
	Die, XTESTEXT1, die!

	* dix/devices.c:
	* include/input.h:
	* mi/mipointer.h:
	* mi/mipointer.c:
	Get rid of horrendous, somewhat misguided, DDX binary-compatibility
	glue.

	* hw/xfree86/common/xf86Config.c:
	* hw/xfree86/common/xf86Configure.c:
	* parser/Module.c:
	* parser/xf86Parser.c:
	Ignore requests to load deprecated GLcore and speedo modules
	(properly, this time).

	* configure.ac:
	* hw/xfree86/common/xf86Globals.c:
	* hw/xfree86/loader/Makefile.am:
	* hw/xfree86/loader/loader.c:
	* hw/xfree86/loader/loader.h:
	* hw/xfree86/loader/loadmod.c:
	* hw/xfree86/loader/ar.h:
	* hw/xfree86/loader/aout.h:
	* hw/xfree86/loader/aoutloader.h:
	* hw/xfree86/loader/aoutloader.c:
	* hw/xfree86/loader/coff.h:
	* hw/xfree86/loader/coffloader.h:
	* hw/xfree86/loader/coffloader.c:
	* hw/xfree86/loader/elf.h:
	* hw/xfree86/loader/elfloader.h:
	* hw/xfree86/loader/elfloader.c:
	* include/xorg-config.h.in:
	* include/xorg-server.h.in:
	Kill the a.out, ELF and COFF loaders with extreme prejudice.  Good
	riddance.  There's more where this came from, too.

	* xkb/xkb.c:
	* xkb/xkbLEDs.c:
	Simplify the 'unsupported' list.

	* configure.ac:
	* Makefile.am:
	* Xext/security.c:
	* dix/colormap.c:
	* dix/devices.c:
	* dix/dixfonts.c:
	* dix/dispatch.c:
	* dix/extension.c:
	* dix/property.c:
	* dix/swaprep.c:
	* include/dix.h:
	* include/dix-config.h:
	* include/dixstruct.h:
	* include/globals.h:
	* include/os.h:
	* include/propertyst.h:
	* include/swaprep.h:
	* lbx/*:
	* mi/miinitext.c:
	* os/Makefile.am:
	* os/access.c:
	* os/connection.c:
	* os/io.c:
	* os/lbxio.c:
	* os/osdep.h:
	* os/utils.c:
	Remove LBX.  Goodbye.

	* hw/xfree86/os-support/bus/linuxPci.c:
	* hw/xfree86/os-support/linux/lnx_pci.c:
	Bug #6583: Only parse /proc/bus/pci/devices once.  (Bill Nottingham)

2006-05-30  Matthieu Herrb <matthieu.herrb@laas.fr>

	* hw/xfree86/os-support/bsd/Makefile.am:
	Powerpc machines also need ioperm_noop.c

2006-05-30  Adam Jackson  <ajax@freedesktop.org>

	* hw/xfree86/doc/man/xorg.conf.man.pre:
	Properly document the DPMS, SyncOnGreen, and TargetRefresh options.

2006-05-29  Kristian Høgsberg  <krh@redhat.com>

	* GL/glx/glxcmds.c: Remove superfluous definition of tfp tokens.

2005-05-24  Daniel Stone  <daniel@freedesktop.org>

	* xkb/ddxKillSrv.c:
	* xkb/ddxPrivate.c:
	* xkb/ddxVT.c:
	Minor #include cleanups.

	* xkb/xkb.c:

hw/darwin/quartz/XDarwinStartup.c
/**************************************************************
 *
 * Startup program for Darwin X servers
 *
 * This program selects the appropriate X server to launch:
 *  XDarwin         IOKit X server (default)
 *  XDarwinQuartz   A soft link to the Quartz X server
 *                  executable (-quartz etc. option)
 *
 * If told to idle, the program will simply pause and not
 * launch any X server. This is to support startx being
 * run by XDarwin.app.
 *
 **************************************************************/
/*
 * Copyright (c) 2001-2002 Torrey T. Lyons. All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * TORREY T. LYONS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
 * OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * Except as contained in this notice, the name of Torrey T. Lyons shall not
 * be used in advertising or otherwise to promote the sale, use or other
 * dealings in this Software without prior written authorization from
 * Torrey T. Lyons.
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/syslimits.h>
#include <ApplicationServices/ApplicationServices.h>

// Macros to build the path name
#ifndef XBINDIR
#define XBINDIR /usr/X11/bin
#endif
#define STR(s) #s
#define XSTRPATH(s) STR(s) "/"
#define XPATH(file) XSTRPATH(XBINDIR) STR(file)

int main(
    int         argc,
    char        *argv[] )
{
    int         i, j, quartzMode = -1;
    char        **newargv;

    // Check if we are going to run in Quartz mode or idle
    // to support startx from the Quartz server. The last
    // parameter in the list is the one used.
    for (i = argc-1; i; i--) {
        if (!strcmp(argv[i], "-idle")) {
            pause();
            return 0;

        } else if (!strcmp(argv[i], "-quartz") ||
                   !strcmp(argv[i], "-rootless") ||
                   !strcmp(argv[i], "-fullscreen"))
        {
            quartzMode = 1;
            break;

        } else if (!strcmp(argv[i], "-iokit")) {
            quartzMode = 0;
            break;
        }
    }

    if (quartzMode == -1) {
#ifdef HAS_CG_MACH_PORT
        // Check if the CoreGraphics window server is running.
        // Mike Paquette says this is the fastest way to determine if it is running.
        CFMachPortRef cgMachPortRef = CGWindowServerCFMachPort();
        if (cgMachPortRef == NULL)
            quartzMode = 0;
        else
            quartzMode = 1;
#else
        // On older systems we assume IOKit mode.
        quartzMode = 0;
#endif
    }

    if (quartzMode) {
        // Launch the X server for the quartz modes

        char quartzPath[PATH_MAX+1];
        int pathLength;
        OSStatus theStatus;
        CFURLRef appURL;
        CFStringRef appPath;
        Boolean success;

        // Build the new argument list
        newargv = (char **) malloc((argc+2) * sizeof(char *));
        for (j = argc; j; j--)
            newargv[j] = argv[j];
        newargv[argc] = "-nostartx";
        newargv[argc+1] = NULL;

        // Use the XDarwinQuartz soft link if it is valid
        pathLength = readlink(XPATH(XDarwinQuartz), quartzPath, PATH_MAX);
        if (pathLength != -1) {
            quartzPath[pathLength] = '\0';
            newargv[0] = quartzPath;
            execv(newargv[0], newargv);
        }

        // Otherwise query LaunchServices for the location of the XDarwin application
        theStatus = LSFindApplicationForInfo(kLSUnknownCreator,
                                             CFSTR("org.x.x11"),
                                             NULL, NULL, &appURL);
        if (theStatus) {
            fprintf(stderr, "Could not find the XDarwin application. (Error = 0x%lx)\n", theStatus);
            fprintf(stderr, "Launch XDarwin once from the Finder.\n");
            return theStatus;
        }

        appPath = CFURLCopyFileSystemPath (appURL, kCFURLPOSIXPathStyle);
        success = CFStringGetCString(appPath, quartzPath, PATH_MAX, CFStringGetSystemEncoding());
        if (! success) {
            fprintf(stderr, "Could not find path to XDarwin application.\n");
            return success;
        }

        // Launch the XDarwin application
        strncat(quartzPath, "/Contents/MacOS/XDarwin", PATH_MAX);
        newargv[0] = quartzPath;
        execv(newargv[0], newargv);
        fprintf(stderr, "Could not start XDarwin application at %s.\n", newargv[0]);
        return errno;

    } else {

        // Build the new argument list
        newargv = (char **) malloc((argc+1) * sizeof(char *));
        for (j = argc; j; j--)
            newargv[j] = argv[j];
        newargv[0] = "XDarwin";
        newargv[argc] = NULL;
    
        // Launch the IOKit X server
        execvp(newargv[0], newargv);
        fprintf(stderr, "Could not start XDarwin IOKit X server.\n");
        return errno;
    }
}
